+precice/@Participant/private/preciceGateway.cpp

```diff
@@ -0,0 +1,354 @@
+                interface->writeData(meshName,dataName,vertexIDs,values);
+                break;
+            }
+            case FunctionID::readData:
+            {
+                const std::string meshName = convertToString(inputs[1]);
+                const std::string dataName = convertToString(inputs[2]);
+                const TypedArray<int32_t> size = inputs[3];
+                const std::vector<int32_t> vertexIDs = convertToIntVector(inputs[4]);
+                size_t dim = interface->getDataDimensions(meshName,dataName);
+                const TypedArray<double> relativeReadTime = inputs[5];
+                std::vector<double> values(size[0]*dim);
+                interface->readData(meshName,dataName,vertexIDs,relativeReadTime[0],values);
+                outputs[0] = factory.createArray<double>({1, size[0]*dim}, values.data(), values.data()+values.size());
+                break;
+            }
+            case FunctionID::requiresGradientDataFor:
+            {
+                const std::string meshName = convertToString(inputs[1]);
+                const std::string dataName = convertToString(inputs[2]);
+                bool requiresGradientData = interface->requiresGradientDataFor(meshName,dataName);
+                outputs[0] = factory.createScalar<bool>(requiresGradientData);
+                break;
+            }
+            case FunctionID::writeGradientData:
+            {
+                const std::string meshName = convertToString(inputs[1]);
+                const std::string dataName = convertToString(inputs[2]);
+                const std::vector<int32_t> vertexIDs = convertToIntVector(inputs[4]);
+                const std::vector<double> gradientValues = convertToDoubleVector(inputs[5]);
+                interface->writeGradientData(meshName,dataName, vertexIDs, gradientValues);
+                break;
+            }
+            case FunctionID::setMeshAccessRegion:
+            {
+                const std::string meshName = convertToString(inputs[1]);
+                const std::vector<double> boundingBox = convertToDoubleVector(inputs[2]);
+                interface->setMeshAccessRegion(meshName,boundingBox);
+                break;
+            }
+            case FunctionID::getMeshVertexIDsAndCoordinates:
+            {
+                const std::string meshName = convertToString(inputs[1]);
+                const TypedArray<int32_t> size = inputs[2];
+                std::vector <int32_t> ids(size[0]);
+                size_t dim = interface->getMeshDimensions(meshName);
+                std::vector <double> positions(size[0]*dim);
+                interface->getMeshVertexIDsAndCoordinates(meshName,ids,positions);
+                outputs[0] = factory.createArray<int32_t>({size[0]}, ids.data(), ids.data()+ids.size());
+                outputs[1] = factory.createArray<double>({size[0], 3}, positions.data(), positions.data()+positions.size());
```

Comment:
```suggestion
                outputs[1] = factory.createArray<double>({size[0], dim}, positions.data(), positions.data()+positions.size());
```
Affects also the layout